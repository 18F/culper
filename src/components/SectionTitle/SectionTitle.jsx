import React from 'react'
import { connect } from 'react-redux'
import { navigation } from '../../config'

/**
 * Provides the section title.
 *
 * Want to update the section title?
 * Using `react-redux` and the `dispatch()` function use the following snippet:
 *
 * ```js
 * this.props.dispatch(updateSection('section', 'subsection', 'Chunky Monkey'))
 * ```
 *
 * `updateSection()` can be found in `SectionActions`
 */
class SectionTitle extends React.Component {
  render () {
    const splitSubsections = (this.props.section.subsection || '').split('/')

    let title = null
    navigation.forEach(s => {
      if (s.url === this.props.section.section) {
        title = trail(breadcrumbs(splitSubsections, s))
      }
    })

    return <h1 className="title">{title}</h1>
  }
}

/**
 * Creates an array of breadcrumbs
 */
const breadcrumbs = (urls, node) => {
  return (node.subsections || []).reduce((a, b) => {
    if (!urls.includes(b.url)) {
      return a
    }

    return a.concat(breadcrumbs(urls, b))
  }, [node.title || node.name])
}

/**
 * Takes an array of breadcrumbs and formats it in to a pretty trail
 */
const trail = (crumbs) => {
  return crumbs.map((crumb, i, arr) => {
    if (arr.length === i + 1) {
      return <span key={`crumb-${i}`} className="title-text">{crumb}</span>
    }

    return <span key={`crumb-${i}`} className="crumb">{crumb} &gt; </span>
  })
}

/**
 * Maps the relevant subtree state from the applications state tree.
 * In this case, we pull the authentication sub-state. This is mapped
 * to the authentication reducer. When actions are dispatched, this
 * method is executed which causes a re-render.
 */
function mapStateToProps (state) {
  const section = state.section || {}
  return {
    section: section
  }
}

// Wraps the the component with connect() which adds the dispatch()
// function to the props property for this component
export default connect(mapStateToProps)(SectionTitle)
