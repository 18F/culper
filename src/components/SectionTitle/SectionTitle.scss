@import 'eqip-colors';
@import 'eqip-fonts';

.eapp-title > .title {
  max-height: 14.5rem;
  overflow: hidden;
  color: #fff;
  font-size: 5rem;
  font-weight: $eapp-bold;

  @media screen and (max-width: 1100px) {
    font-size: 4rem;
  }

  > .title-text {
    display: block;
    line-height: 1.2;
  }
}
