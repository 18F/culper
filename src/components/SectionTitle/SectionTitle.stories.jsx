import React from 'react'
import { storiesOf } from '@storybook/react'

import SectionTitle from './SectionTitle'

storiesOf('SectionTitle', module)
  .add('default', () => (
    <SectionTitle />
  ))
