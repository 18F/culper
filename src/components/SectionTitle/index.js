import SectionTitle from './SectionTitle'
export { SectionTitle }
