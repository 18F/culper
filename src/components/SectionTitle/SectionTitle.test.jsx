import React from 'react'
import MockAdapter from 'axios-mock-adapter'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { Provider } from 'react-redux'
import SectionTitle from './SectionTitle'
import { mount } from 'enzyme'

describe('The title section', () => {
  // Setup
  const middlewares = [ thunk ]
  const mockStore = configureMockStore(middlewares)

  it('can handle no title', () => {
    const store = mockStore({ section: { section: '' } })
    const component = mount(<Provider store={store}><SectionTitle /></Provider>)
    expect(component.find('h1').length).toEqual(1)
    expect(component.find('h1').text()).toEqual('')
  })

  it('can handle title 1 deep', () => {
    const store = mockStore({ section: { section: 'legal', subsection: '' } })
    const component = mount(<Provider store={store}><SectionTitle /></Provider>)
    expect(component.find('h1').length).toEqual(1)
    expect(component.find('h1').text()).toEqual('Investigative and criminal history')
  })

  it('can handle title 2 deep', () => {
    const store = mockStore({ section: { section: 'legal', subsection: 'technology' } })
    const component = mount(<Provider store={store}><SectionTitle /></Provider>)
    expect(component.find('h1').length).toEqual(1)
    expect(component.find('h1').text()).toEqual('Investigative and criminal history > Use of information technology systems')
  })

  it('can handle title 3 deep', () => {
    const store = mockStore({ section: { section: 'legal', subsection: 'technology/unauthorized' } })
    const component = mount(<Provider store={store}><SectionTitle /></Provider>)
    expect(component.find('h1').length).toEqual(1)
    expect(component.find('h1').text()).toEqual('Investigative and criminal history > Use of information technology systems > Unauthorized access')
  })
})
