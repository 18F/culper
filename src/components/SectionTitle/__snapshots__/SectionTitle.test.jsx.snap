// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The SectionTitle component renders without errors 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": null,
  "nodes": Array [
    null,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <SectionTitle
        location={Object {}}
      />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "location": Object {},
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": null,
        "_debugID": 2,
        "_renderedOutput": null,
      },
      "_renderedNodeType": 2,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <SectionTitle
    location={Object {}}
  />,
}
`;
